anomalib/data/base.py

```diff
@@ -4,33 +4,88 @@
+from pathlib import Path
 from typing import Dict, Optional, Tuple, Union
 
 import albumentations as A
-import cv2
 import numpy as np
+import pandas as pd
 from pandas import DataFrame
 from pytorch_lightning import LightningDataModule
+from pytorch_lightning.trainer.states import TrainerFn
 from pytorch_lightning.utilities.types import EVAL_DATALOADERS, TRAIN_DATALOADERS
 from torch import Tensor
 from torch.utils.data import DataLoader, Dataset
 
-from anomalib.data.utils import read_image
+from anomalib.data.utils import read_image, read_mask
 from anomalib.pre_processing import PreProcessor
 
 logger = logging.getLogger(__name__)
 
+TASK_CLASSIFICATION = "classification"
+TASK_SEGMENTATION = "segmentation"
+TASKS = (TASK_CLASSIFICATION, TASK_SEGMENTATION)
+
+SPLIT_TRAIN = "train"
+SPLIT_VAL = "val"
+SPLIT_TEST = "test"
+SPLITS = (SPLIT_TRAIN, SPLIT_VAL, SPLIT_TEST)
+
+LABEL_NORMAL = 0
+LABEL_ANOMALOUS = 1
+
+_SAMPLES_MANDATORY_COLUMNS_CLASSIFICATION = ["image_path", "label"]
+_SAMPLES_MANDATORY_COLUMNS_SEGMENTATION = _SAMPLES_MANDATORY_COLUMNS_CLASSIFICATION + ["mask_path"]
+
 
 class AnomalibDataset(Dataset):
     """Anomalib dataset."""
 
-    def __init__(self, samples: DataFrame, task: str, split: str, pre_process: PreProcessor):
+    @staticmethod
+    @abstractmethod
+    def _create_samples(root: Path, split: str) -> DataFrame:
+        """This method should be implemented in the subclass.
+
+        This method should return a dataframe that contains the information needed by __getitem__ to load each of
+        the dataset items into memory. The dataframe must at least contain the following columns:
+            image_path - Path to file system location where the image is stored.
+            label - Index of the anomaly label, typically 0 for "normal" and 1 for "anomalous".
+
+        Additionally, when the task type is segmentation, the dataframe must have the mask_path column, which contains
+        the path the ground truth masks (for the anomalous images only).
+
+        Example of a dataframe returned by calling this method from a concrete class:
+        |---|-------------------|-----------|-------------|------------------|
+        |   | image_path        | label_str | label       | mask_path        |
+        |---|-------------------|-----------|-------------|------------------|
+        | 0 | path/to/image.png | anomalous | 0           | path/to/mask.png |
+        |---|-------------------|-----------|-------------|------------------|
+        """
+        pass
+
+    def __init__(self, root: Path, task: str, split: str, pre_process: PreProcessor):
         super().__init__()
-        self.samples = samples
+
+        assert split in SPLITS, f"Unknown split: {split}"
+        assert task in TASKS, f"Unknown task: {task}"
+
+        self.root = root
         self.task = task
         self.split = split
         self.pre_process = pre_process
+        self.samples: DataFrame = self._create_samples(self.root, self.split)
```

Comment:
The main problem with having `_create_samples` as an abstract method of AnomalibDataset, is that this method involves randomly assigning the samples to one of the subsets (train, val, test). This means that we would need to pass a seed to the dataset instance to make sure that the samples in each of the subsets remain mutually exclusive. This also means that we would generate the exact same samples table in each call of `_create_samples`, which would be inefficient.

The more sensible approach, therefore, would be to create the samples once in a common location, then apply the subset splitting, and then only pass the samples of each split to the constructors of the respective dataset instances. In line with this reasoning, we moved `_create_samples` to the datamodule.

I do agree that it would be better to decouple `AnomalibDataModule` and `AnomalibDataset` to make the dataset more standalone. I'm also not very happy that we're passing a dataframe around between classes, so we might need yet another design for this.

Reply:
hmm.. ok i see, good point

Reply:
I don't how fast you want to get over with this pr, but here is a possible alternative:

AnomalibDataset is not aware of these dynamically defined splits and deals with dataset-defines splits (mvtec would have train and test only, mvtec loco has train, val, and test)

AnomalibDataset gets still gets a df at the init like your version but should provide an abstract class method to load the splits mentioned above.

AnomalivDataset has an interface or method to generate subsplits (get two instances from one such that they dont have common samples)

this interface could take two lists of indices so the seed that generates them stays out of this function 

(Probably better) alternative: the list of indices for subsampling can be given at the abstract classmethod mentioned before

AnomalibDatamodule would deal with the subspliting at the init, by only generating the lists of indices from above (must know the nb of samples in advance but no big deal)

Reply:
If you want, let's make a call to discuss

Reply:
@jpcbertoldo I've outlined a new design proposal in https://github.com/openvinotoolkit/anomalib/pull/572. We could continue the discussion there. I will close this PR now.